### ISSUE TITLE
System.register does not follow correct normalization rules.

### ISSUE THREAD

**Issue by Zoltu** [systemjs/systemjs#602]
Originally troubleshooted here: https://github.com/jspm/jspm-cli/issues/912

```html
<!doctype html>
<html>
	<head></head>
	<body>
		<script src="jspm_packages/system.js"></script>
		<script>
			System.config({
				"baseURL": "/",
				"defaultJSExtensions": true,
				"packages": {
					"source": {
						"defaultExtension": "ts"
					}
				}
			});
			System.register('source/main.ts', [], function(x) {});
			console.log(Object.keys(System.defined)[0]);
			System.normalize('source/main').then(function(x) { console.log(x); });
		</script>
	</body>
</html>
```
The above will print the following:
```
http://localhost:8080/source/main.ts.js
http://localhost:8080/source/main.ts
```

`System.register` is not following the full set of normalization rules.  In this case, the `defaultExtension` should be `ts`, rather than `js` since the module is inside the `source` package.

This is particularly problematic because the above code is akin to what is output by JSPM bundle when bundling `.ts` files.  This results in TypeScript not being loaded from bundles.

A related bug is that `System.register` will re-normalized full URIs.  For example, if you change the above call to:
```js
System.register('http://localhost:8080/source/main.ts', [], function(x) {});
```
You will get the same output, once again the `.js` suffix is added even though the name is already fully normalized into a URI.

**Comment by guybedford** [comment]
Thanks, yes this is a bug.

**Comment by guybedford** [comment]
Does this work if you disable `defaultJSExtensions` for the built version of the page? That may be advisable here to work around the issue as a full fix is quite tricky.

Tempted to rather focus efforts on getting the jspm 0.17 beta out sooner which will avoid these issues when we can forget the `defaultJSExtensions` flag behind.

Let me know your thoughts though.

**Comment by Zoltu** [comment]
Disabling `defaultJSExtensions` breaks Aurelia from what I can tell, which is the framework I'm using for my actual project.  Is it possible to make Aurelia work with `defaultJSExtension` disabled?  I dabbled for about 30 minutes when JSPM 0.16 came out to try and get it working but was unable to.  I *want* to switch to TS by default, but it needs to work with third party libraries that are written/distributed as JS first.  Is that what JSPM 0.17 will allow for?

**Comment by guybedford** [comment]
I mean specifically disabling defaultJSExtensions in the config when everything is in a bundle.

**Comment by Zoltu** [comment]
With `defaultJSExtensions: true`
`jspm bundle source/main + aurelia-bootstrapper`

Then with `defaultJSExtensions: false` I launch the app and it fails to load aurelia-bootstrapper.  I do see `define("github:aurelia/bootstrapper@0.14.0/index.js"` in my `build.js`, implying that it was bundled... it just doesn't get loaded from the bundle for whatever reason so it tries to load from the server which fails.

Interestingly, when I switch back to `defaultJSExtensions: true`, `aurelia-bootstrapper` isn't fetched from the server.  This implies to me that having `defaultJSExtensions: true` during bundle and `false` during run doesn't work.

Note: I really am floundering here, I don't fully understand how all of the default extension stuff works, I have been waiting until 0.17 to really understand since as I understand it, everything is about to change.

**Comment by guybedford** [comment]
When you say it fails to load, what sort of error do you see?

Yes this is simply a bug in builds for `defaultExtension` interacting with `defaultJSExtensions`. It's not that everything is going to change, it's just that `defaultJSExtensions` is a stop gap, that does cause some difficult configuration scenarios like this one.

**Comment by Zoltu** [comment]
Sorry, I thought I pasted in the error.  I receive a 404 when attempting to `GET` `http://localhost:8080/jspm_packages/github/aurelia/bootstrapper@0.14.0`.  As expected, there is no `.js` suffix.

Looking at the debugger, I can see `http://localhost:8080/jspm_packages/github/aurelia/bootstrapper@0.14.0.js` in the list of modules brought in via the bundle, but when the call to `System.import('aurelia-bootstrapper')` resolves the module ID it ends up looking for `"http://localhost:8080/jspm_packages/github/aurelia/bootstrapper@0.14.0"`, which it doesn't find.

With JSPM 0.17, is it expected that I'll be able to depend on third party JS libraries (like Aurelia) with my project being TS?  Will Aurelia need to change anything on their end to make this work or will JSPM be able to figure it all out automatically (for example, on `jspm install`)?

I don't mind waiting for 0.17 if you believe it will allow me to bundle TypeScript correctly while still leveraging JS third party dependencies.

**Comment by guybedford** [comment]
@Zoltu right the workaround technique it would only work if every single module in your app was loaded from an existing bundle - any individual module requests would be incorrect as they don't have extensions.

Unfortunately jspm 0.17 will be a little while away so I wouldn't suggest waiting for it.

Let me see if I can add a workaround to at least avoid any extension-adding for bundle normalization.

**Comment by Zoltu** [comment]
Perhaps I am misunderstanding, but when I bundled I did bundle everything (with default extension turned on).  Then at runtime I turned default extension off and SystemJS didn't use the files from the bundle.  See https://github.com/systemjs/systemjs/issues/602#issuecomment-124558998.

I suspect that you have a much stronger understanding of the problem at this point so I'll trust you to figure out the right way to deal with the problem.  My desire *as a user* is to be able to bundle a project full of `.ts` files that has `.js` dependencies and not have to ship the TypeScript compiler to the browser.

**Comment by guybedford** [comment]
@Zoltu did you try including the bundle with a script tag after SystemJS? Just trying to suggest a way forward for you here.

Will certainly see what I can do this side to fix the issue.

**Comment by Zoltu** [comment]
My real project looks something like this:
```html
<script src="jspm_packages/system.js"></script>
<script src="jspm-config.js"></script>
<script src="build.js"></script>
<script>
	System.import('aurelia-bootstrapper');
</script>
```

Where `build.js` is the `jspm bundle` output.  That is what I have done most of my real testing with (outside of the repro case in my original issue post).  Is that what you meant by including the bundle after SystemJS?  Is there some other way to include a bundle besides this?

**Comment by guybedford** [comment]
Try kicking off your first import with `System.import('github/aurelia/bootstrapper@0.14.0.js')`?

**Comment by guybedford** [comment]
Sorry `github:aurelia/bootstrapper@0.14.0.js`

**Comment by Zoltu** [comment]
That got me further in that `aurelia-bootstrapper` appears to have been loaded from the bundle, but then I received a 404 trying to download `http://localhost:8080/jspm_packages/github/aurelia/loader-default@0.9.0`.  This is with `defaultJSExtension: false` in the runtime config, true in the bundle config.

**Comment by guybedford** [comment]
@Zoltu I've just added a test for this in https://github.com/systemjs/systemjs/commit/039134c947134c3c313764194bada0079b43cae1 and it is definitely working ok for me. Perhaps try running the SystemJS master version (`jspm dl-loader --edge`)? If you can suggest a change to the test to get it to fail that would be appreciated too.

**Comment by Zoltu** [comment]
Full repro steps (see also my comment on the commit: https://github.com/systemjs/systemjs/commit/039134c947134c3c313764194bada0079b43cae1#commitcomment-12411907)
```
> mkdir test
> cd test
> npm install jspm@beta
> node_modules/.bin/jspm dl-loader edge
# defaults for all prompts except TypeScript for transpiler
> npm install node-static
> cat > index.html
<!doctype html>
<html>
	<head></head>
	<body>
		<script src="jspm_packages/system.js"></script>
		<script>
			System.config({
				"defaultJSExtensions": true,
				"packages": {
					"foo": {
						"defaultExtension": "ext"
					}
				}
			});
			System.register('foo/custom.ext', [], function(_export) {
				return {
					setters: [],
					execute: function() {
						_export('zip', 'zap');
					}
				};
			});
			System.import('foo/custom').then(function(m) {
				console.log(m.zip);
			});
		</script>
	</body>
</html>
> node_modules/.bin/static
# browse to localhost:8080
# note error in browser console and 404 when trying to download foo/custom.ext
```

**Comment by guybedford** [comment]
Thanks, I've included a fix now.

**Comment by guybedford** [comment]
Released in 0.18.5.
